sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureDirectoryStream.java

```diff
@@ -0,0 +1,188 @@
+        private final DirectoryStream.Filter<? super Path> filter;
+        private final Iterator<BlobItem> blobIterator;
+        private final AzurePath path;
+        private Path bufferedNext = null;
+        private final Set<String> directoryPaths;
+
+        AzureDirectoryIterator(AzureDirectoryStream parentStream, AzurePath path,
+            DirectoryStream.Filter<? super Path> filter) throws IOException {
+            this.parentStream = parentStream;
+            this.filter = filter;
+            this.path = path;
+            directoryPaths = new HashSet<>();
+
+            BlobContainerClient containerClient;
+            ListBlobsOptions listOptions = new ListBlobsOptions()
+                .setDetails(new BlobListDetails().setRetrieveMetadata(true));
+            if (path.isRoot()) {
+                String containerName = path.toString().substring(0, path.toString().length() - 1);
+                containerClient = ((AzureFileSystem) path.getFileSystem()).getBlobServiceClient()
+                    .getBlobContainerClient(containerName);
+            } else {
+                AzureResource azureResource = new AzureResource(path);
+                listOptions.setPrefix(azureResource.getBlobClient().getBlobName() + AzureFileSystem.PATH_SEPARATOR);
+                containerClient = azureResource.getContainerClient();
+            }
+            this.blobIterator = containerClient
+                .listBlobsByHierarchy(AzureFileSystem.PATH_SEPARATOR, listOptions, null).iterator();
+        }
+
+        @Override
+        public boolean hasNext() {
+            // Closing the parent stream halts iteration.
+            if (parentStream.closed) {
+                return false;
+            }
+
+            // In case a customer calls hasNext multiple times in a row. If we've buffered an element, we have a next.
+            if (this.bufferedNext != null) {
+                return true;
+            }
+
+            /*
+            Search for a new element that passes the filter and buffer it when found. If no such element is found,
+            return false.
+             */
+            while (this.blobIterator.hasNext()) {
+                BlobItem nextBlob = this.blobIterator.next();
+                Path nextPath = getNextListResult(nextBlob);
+                try {
+                    if (this.filter.accept(nextPath) && passesDirectoryDuplicateFilter(nextPath, nextBlob)) {
+                        this.bufferedNext = nextPath;
+                        return true;
+                    }
+                } catch (IOException e) {
+                    throw LoggingUtility.logError(logger, new DirectoryIteratorException(e));
+                }
+            }
+            return false;
+        }
+
+        @Override
+        public Path next() {
+            if (this.bufferedNext == null) {
+                if (!this.hasNext()) { // This will populate cachedNext in the process.
```

Comment:
should these comments actually be bufferedNext? I'm assuming you just renamed the variable and forgot to change the comments